signer-service/src/api/services/siwe.service.js

```diff
@@ -0,0 +1,74 @@
+const siwe = require('siwe');
+const { createPublicClient, http } = require('viem');
+const { polygon } = require('viem/chains');
+
+// Make constants on config
+const scheme = 'https';
+const domain = 'satoshipay.io';
+const origin = 'https://vortex.com';
+const statement = 'Please sign the message to login!';
+
+// Map that will hold the siwe messages sent + nonce we defined
+const siweMessagesMap = new Map();
+
+exports.createAndSendSiweMessage = async (address) => {
+  const nonce = siwe.generateNonce();
+  const siweMessage = new siwe.SiweMessage({
+    scheme,
+    domain,
+    address,
+    statement,
+    uri: origin,
+    version: '1',
+    chainId: polygon.id,
+    nonce,
+    expirationTime: new Date(Date.now() + 360 * 60 * 1000).toISOString(),
+  });
+  const preparedMessage = siweMessage.toMessage();
+  siweMessagesMap.set(nonce, { siweMessage, address });
+
+  return { siweMessage: preparedMessage, nonce };
+};
+
+exports.verifySiweMessage = async (nonce, signature) => {
+  const maybeSiweData = siweMessagesMap.get(nonce);
+  if (!maybeSiweData) {
+    throw new Error('Message not found, we have not send this message or nonce is incorrect.');
+  }
+
+  const publicClient = createPublicClient({
+    chain: polygon,
+    transport: http(),
+  });
+
+  const valid = await publicClient.verifyMessage({
+    address: maybeSiweData.address,
+    message: maybeSiweData.siweMessage.toMessage(),
+    signature,
+  });
+
+  if (!valid) {
+    throw new Error('Invalid signature.');
+  }
+
+  // Perform additional checks to ensure message integrity
+  if (maybeSiweData.siweMessage.nonce !== nonce) {
+    throw new Error('Nonce mismatch.');
+  }
+
+  if (maybeSiweData.expirationTime && new Date(maybeSiweData.expirationTime) < new Date()) {
+    throw new Error('Message has expired.');
+  }
+
+  // Successful verification, remove messages with same address
+  // from map except for the one we just verified.
+  siweMessagesMap.forEach((data, nonce) => {
+    if (data.address === maybeSiweData.address && nonce !== maybeSiweData.siweMessage.nonce) {
+      siweMessagesMap.delete(nonce);
+    }
+  });
+
+  return maybeSiweData.siweMessage;
+};
+
+// TODO we need some sort of session log-out.
```

Comment:
It will grow indefinitely for new addresses. This simple way of keeping valid sessions is very rudimentary, that's why I would use a session library if we want more functionality.

Regarding if we need it or not, technically no, but  I believe it's just a standard. For instance [polkassembly](https://pendulum.polkassembly.io/) has a log out "feature". Again we need to define this.

Reply:
Yeah maybe we should use sessions. Or at least persist the nonce map so that when the server restarts (due to redeployment or crash) users don't need to always sign in again which would be the case if we only keep an in-memory list.

Reply:
I think that adding more complexity to this implementation, for instance the persistence after a crash, comes very close to a full session implementation which I would prefer to do with a library already.

I'm even dubious about the logout. Maybe we can skip it for now wdyt? And in the next iteration we use a proper session library.

Reply:
I agree and also would suggest we skip a logout for now. Doesn't add much value at this point.